src/webgpu/shader/execution/expression/call/builtin/textureDimensions.spec.ts

```diff
@@ -0,0 +1,497 @@
+export const description = `
+Execution tests for the 'textureDimensions' builtin function
+
+The dimensions of the texture in texels.
+For textures based on cubes, the results are the dimensions of each face of the cube.
+Cube faces are square, so the x and y components of the result are equal.
+If level is outside the range [0, textureNumLevels(t)) then any valid value for the return type may be returned.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import {
+  isCompressedTextureFormat,
+  kAllTextureFormats,
+  kColorTextureFormats,
+  kTextureFormatInfo,
+} from '../../../../../format_info.js';
+import { GPUTest } from '../../../../../gpu_test.js';
+
+export const g = makeTestGroup(GPUTest);
+
+/** @returns the aspects to test for the given format */
+function aspectsForFormat(format: GPUTextureFormat): readonly GPUTextureAspect[] {
+  switch (format) {
+    case 'depth24plus-stencil8':
+    case 'depth32float-stencil8':
+      return ['depth-only', 'stencil-only'];
+    default:
+      return ['all'];
+  }
+}
+
+/** @returns the sample counts to test for the given format */
+function samplesForFormat(format: GPUTextureFormat): readonly number[] {
+  const info = kTextureFormatInfo[format];
+  return info.multisample ? [1, 4] : [1];
+}
+
+/**
+ * @returns the number of texture mipmap levels to test for the format, dimensions and number of
+ * samples.
+ */
+function textureMipCount(params: {
+  format: GPUTextureFormat;
+  dimensions: GPUTextureViewDimension;
+  samples?: number;
+}): readonly number[] {
+  if (params.dimensions === '1d') {
+    return [1];
+  }
+  if (params.samples !== undefined && params.samples !== 1) {
+    // https://www.w3.org/TR/webgpu/#texture-creation
+    // If descriptor.sampleCount > 1: descriptor.mipLevelCount must be 1.
+    return [1];
+  }
+  const info = kTextureFormatInfo[params.format];
+  if (
+    isCompressedTextureFormat(params.format) ||
+    info.depth !== undefined ||
+    info.stencil !== undefined
+  ) {
+    // https://www.w3.org/TR/webgpu/#dom-gputexturedimension-2d
+    // Only "2d" textures may have mipmaps, be multisampled, use a compressed or depth/stencil
+    // format, and be used as a render attachment.
+    return [1];
+  }
+  return [1, 3];
+}
+
+/** @returns the baseMipLevel on the texture view to test */
+function baseMipLevel(params: { textureMipCount: number }): readonly number[] {
+  const out: number[] = [];
```

Comment:
> How about
> 
> ```js
> return [...Array(params.textureMipCount).keys()]
> ```

That's a GC-heavy way to produce a sequence of numbers.

> Or make a helper `range(k)` that returns the list `[0,..,k-1]` and call that. For added protection, it can assert that k is a non-negative integer. (And the list is empty if k = 0)

Could do, but this is 4 lines of KISS code, which is easily adaptable (like to reduce permutations if needed). I see little reason to add helpers for something so trivial.